include/fmt/format.h

```diff
@@ -1822,83 +1844,208 @@ class specs_handler: public specs_setter<typename Context::char_type> {
-  };
+    Name name;  // This is not string_view because of gcc 4.4.
+  } val;
 };
 
 // Format specifiers with width and precision resolved at formatting rather
 // than parsing time to allow re-using the same parsed specifiers with
 // differents sets of arguments (precompilation of format strings).
 template <typename Char>
 struct dynamic_format_specs : basic_format_specs<Char> {
-  arg_ref<Char> width_ref;
-  arg_ref<Char> precision_ref;
+  typedef arg_ref<Char, string_value<Char>> arg_reference;
+  arg_reference width_ref;
+  arg_reference precision_ref;
 };
 
 // Format spec handler that saves references to arguments representing dynamic
 // width and precision to be resolved at formatting time.
-template <typename ParseContext>
-class dynamic_specs_handler :
-    public specs_setter<typename ParseContext::char_type> {
+template <typename Specs, typename ParseContext, typename ArgRefCreator>
+class dynamic_specs_handler
+    : public specs_setter<typename ParseContext::char_type> {
  public:
   typedef typename ParseContext::char_type char_type;
 
-  FMT_CONSTEXPR dynamic_specs_handler(
-      dynamic_format_specs<char_type> &specs, ParseContext &ctx)
-    : specs_setter<char_type>(specs), specs_(specs), context_(ctx) {}
+  FMT_CONSTEXPR dynamic_specs_handler(Specs &specs, ParseContext &ctx,
+                                      ArgRefCreator &creator)
+      : specs_setter<char_type>(specs),
+        specs_(specs),
+        context_(ctx),
+        arg_ref_creator_(creator) {}
 
   FMT_CONSTEXPR dynamic_specs_handler(const dynamic_specs_handler &other)
-    : specs_setter<char_type>(other),
-      specs_(other.specs_), context_(other.context_) {}
+      : specs_setter<char_type>(other),
+        specs_(other.specs_),
+        context_(other.context_),
+        arg_ref_creator_(other.arg_ref_creator_) {}
 
   template <typename Id>
   FMT_CONSTEXPR void on_dynamic_width(Id arg_id) {
-    specs_.width_ref = make_arg_ref(arg_id);
+    specs_.width_ref = arg_ref_creator_.create(arg_id);
   }
 
   template <typename Id>
   FMT_CONSTEXPR void on_dynamic_precision(Id arg_id) {
-    specs_.precision_ref = make_arg_ref(arg_id);
+    specs_.precision_ref = arg_ref_creator_.create(arg_id);
   }
 
   FMT_CONSTEXPR void on_error(const char *message) {
     context_.on_error(message);
   }
 
  private:
-  typedef arg_ref<char_type> arg_ref_type;
+  Specs &specs_;
+  ParseContext &context_;
+  ArgRefCreator &arg_ref_creator_;
+};
 
-  template <typename Id>
-  FMT_CONSTEXPR arg_ref_type make_arg_ref(Id arg_id) {
+template <typename ParseContext>
+class dynamic_arg_ref_creator {
+  typedef typename ParseContext::char_type char_type;
+
+ public:
+  typedef arg_ref<char_type, string_value<char_type>> arg_ref_type;
+
+  FMT_CONSTEXPR dynamic_arg_ref_creator(ParseContext &ctx) : context_(ctx) {}
+
+  FMT_CONSTEXPR arg_ref_type create(unsigned arg_id) {
     context_.check_arg_id(arg_id);
```

Comment:
I think this belongs to `dynamic_specs_handler`. Same with the calls to `check_arg_id` and `next_arg_id` below. Then this class won't need to store a reference to parse context.

Reply:
You're right, I'll move `check_arg_id` back to the `dynamic_specs_handler`, but the context reference is still needed to get the id when `create(auto_id)` is called.

Reply:
Moved checking argument id back to the `dynamic_specs_handler`.